modules/junit-jupiter/src/test/java/org/testcontainers/junit/jupiter/TestLifecycleAwareSignallingTest.java

```diff
@@ -0,0 +1,56 @@
+package org.testcontainers.junit.jupiter;
+
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.MethodOrderer.OrderAnnotation;
+import org.junit.jupiter.api.Order;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.TestMethodOrder;
+import org.junit.jupiter.api.extension.AfterAllCallback;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.junit.jupiter.api.extension.ExtensionContext;
+
+import static org.junit.Assert.assertEquals;
+
+// The order of @ExtendsWith and @Testcontainers is crucial in order for the static case to be tested
+@ExtendWith({TestLifecycleAwareSignallingTest.SharedContainerAfterAllTestExtension.class})
+@Testcontainers
+@TestMethodOrder(OrderAnnotation.class)
+class TestLifecycleAwareSignallingTest {
```

Comment:
Is there anything we could do to test:
* correct ordering (e.g. that after doesn't happen before before)
* correct passing of a throwable into `afterTest` when a failure occurs

These are both relatively minor when looking at the code (it's clear!) but it would be good to have as protection against any future regressions.

Reply:
I'm feeling a bit dirty writing this hacky tests but I think that I've ben able to test ordering and exception captured by non-shared testcontainers.

Reply:
Thanks for making an effort to get this covered! I think we could perhaps make this a little less uncomfortable - tests that rely on timing/sleeps always make me nervous, even when every effort has been made to make it safe!

What if we modify the `afterTest` method in the mock:

```java
    @Override
    public void afterTest(TestDescription description, Optional<Throwable> throwable) {
        numAfterTestsCalls++;
        if (numAfterTestsCalls > numBeforeTestsCalls) {
            throw new IllegalStateException("afterTest seems to have been called without a corresponding beforeTest");
        }
    }
````

This way, we don't need to know exactly when in time the methods were called, but we still can be confident about the order.

WDYT?

Reply:
This is an interesting idea. I also don't like to have tests being time dependent.
However in this case it is all about timing (`beforeTest()` should be called before `afterTest()`).
Also this would mean that a documenting test case would become obsolete. While this saves one test do be executed it basically moves the test into the mock.
Since both approaches have their upsides and downsides I would be fine to change the test as you suggested.
WDYT?

Reply:
One can also record a history of invocations (e.g. `List<String>`) and assert it

Reply:
I like that idea, thanks! I'll give it a try.

Reply:
@rnorth I've changed the tests as recommended by @bsideup . I've also separated the tests into two test classes. One for the lifecycle method themselves and one for the exception capturing. I know that this means that another container has to be started but I like that the tests have now become very clear.

Reply:
Awesome, looks good to me. Thanks for the alternative suggestion @bsideup!